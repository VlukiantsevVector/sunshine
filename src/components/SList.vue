<!--
  Example: <s-list :items="items" v-model="selected" :filter="keyword" checkable/>

 items     - The items to show in the list
 v-model   - The items in the list that is selected/checked.
             If some items are to be preselected, they have to
             be added to items to be visible in the list
 filter    - Only show items containing the filter keyword
-->

<template>
  <ul
    tabIndex="0"
    :class="classes()"
  >
    <s-list-item
      v-for="item in visibleItems"
      :key="item.key"
      :selected="isSelected(item)"
      @click.exact="onItemClick($event, item)"
      @click.ctrl.exact="onItemCtrlClick(item)"
      @click.shift.exact="onItemShiftClick(item)"
    >
      {{ item.label }}
    </s-list-item>
  </ul>
</template>

<script>
import Vue from 'vue';
import mixins from 'vue-typed-mixins';
import { ClassesMixin } from '../lib/utils';
import SListItem from './SListItem.vue';

export default mixins(ClassesMixin).extend({
  name: 's-list',

  components: {
    SListItem,
  },

  props: {
    items: {
      type: Array,
      required: true,
    },

    value: {
      type: Array,
      required: true,
    },

    checkable: {
      type: Boolean,
      default: false,
    },

    filter: {
      type: String,
      default: '',
    },
  },

  data() {
    return {
      internalValue: [],
    };
  },

  watch: {
    value: {
      handler(val) {
        this.internalValue = val;
      },
      immediate: true,
    },

    internalValue(val) {
      if (val !== this.value) {
        this.$emit('input', val);
      }
    },
  },

  computed: {
    visibleItems() {
      const filter = this.filter.toLowerCase();
      return this.items.filter(i => i.label.toLowerCase().includes(filter));
    },
  },

  methods: {
    isSelected(item) {
      return this.internalValue.map(i => i.key).includes(item.key);
    },

    onItemClick(event, item) {
      if (event.detail > 1) {
        this.$emit('dblclick', item);
      } else {
        this.internalValue = this.isSelected(item) ? [] : [item];
      }
    },

    onItemCtrlClick(item) {
      this.internalValue = this.isSelected(item) ?
        this.internalValue.filter(i => i.key !== item.key) :
        this.internalValue.concat(item);
    },

    onItemShiftClick(item) {
      // implement
    },
  },
});

</script>
