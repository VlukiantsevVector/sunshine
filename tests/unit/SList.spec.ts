import { mount } from '@vue/test-utils';
import SList from '@/components/SList.vue';

const COUNTRIES = [
  'Argentina',
  'Belgium',
  'Canada',
  'Denmark',
  'France',
  'Germany',
];

const SELECTED = [
  'Belgium',
  'Canada',
];

let items : Array<{}> = [];
let value : Array<String> = [];

beforeAll(() => {
  items = COUNTRIES.map(c => ({ key: c.toLowerCase(), label: c }));
  value = SELECTED.map(c => c.toLowerCase());
});

describe('SList', () => {
  it('renders empty list', () => {
    const wrapper = mount(SList, {
      propsData: {
        items: [],
        value: [],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders list', () => {
    const wrapper = mount(SList, {
      propsData: {
        items,
        value: [],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders list with preselected items', () => {
    const wrapper = mount(SList, {
      propsData: {
        items,
        value,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders list with selected item by click', () => {
    const wrapper = mount(SList, {
      propsData: {
        items,
        value: [],
      },
    });

    const item = wrapper.find('.s-list-item');
    item.element.click();
    expect(wrapper.html()).toMatchSnapshot();
  });


  it('emits selected values', () => {
    const wrapper = mount(SList, {
      propsData: {
        items,
        value,
      },
    });
    const item = wrapper.find('.s-list-item');
    item.element.click();
    expect(wrapper.emitted('input')[0][0]).toEqual(['argentina']);
  });

  it('emits all selected values on ctrl click', () => {
    const wrapper = mount(SList, {
      propsData: {
        items,
        value,
      },
    });
    const item = wrapper.find('.s-list-item');
    item.trigger('click', { ctrlKey: true });
    expect(wrapper.emitted('input')[0][0]).toEqual(['belgium', 'canada', 'argentina']);
  });
});
