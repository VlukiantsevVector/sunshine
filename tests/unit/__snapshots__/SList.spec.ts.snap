// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SList renders empty list 1`] = `<ul tabindex="0" class="s-list"></ul>`;

exports[`SList renders list 1`] = `
<ul tabindex="0" class="s-list">
  <li class="s-list-item">
    <!---->
    Argentina
  </li>
  <li class="s-list-item">
    <!---->
    Belgium
  </li>
  <li class="s-list-item">
    <!---->
    Canada
  </li>
  <li class="s-list-item">
    <!---->
    Denmark
  </li>
  <li class="s-list-item">
    <!---->
    France
  </li>
  <li class="s-list-item">
    <!---->
    Germany
  </li>
</ul>
`;

exports[`SList renders list with preselected items 1`] = `
<ul tabindex="0" class="s-list">
  <li class="s-list-item">
    <!---->
    Argentina
  </li>
  <li class="s-list-item s-list-item--selected">
    <!---->
    Belgium
  </li>
  <li class="s-list-item s-list-item--selected">
    <!---->
    Canada
  </li>
  <li class="s-list-item">
    <!---->
    Denmark
  </li>
  <li class="s-list-item">
    <!---->
    France
  </li>
  <li class="s-list-item">
    <!---->
    Germany
  </li>
</ul>
`;

exports[`SList renders list with selected item by click 1`] = `
<ul tabindex="0" class="s-list">
  <li class="s-list-item s-list-item--selected">
    <!---->
    Argentina
  </li>
  <li class="s-list-item">
    <!---->
    Belgium
  </li>
  <li class="s-list-item">
    <!---->
    Canada
  </li>
  <li class="s-list-item">
    <!---->
    Denmark
  </li>
  <li class="s-list-item">
    <!---->
    France
  </li>
  <li class="s-list-item">
    <!---->
    Germany
  </li>
</ul>
`;
